src/btree/bt_ret.c

```diff
@@ -142,8 +142,20 @@ __value_return(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd)
 int
 __wt_key_return(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
 {
-	WT_RET(__key_return(session, cbt));
+	WT_CURSOR *cursor;
+
+	cursor = &cbt->iface;
 
+	/*
+	 * We may already have an internal key, in which case the cursor may
+	 * not be set up to get another copy (for example, when we rely on a
+	 * search-function result).
+	 */
+	F_CLR(cursor, WT_CURSTD_KEY_EXT);
+	if (!F_ISSET(cursor, WT_CURSTD_KEY_INT)) {
+		WT_RET(__key_return(session, cbt));
+		F_SET(cursor, WT_CURSTD_KEY_INT);
+	}
```

Comment:
If you call `WT_CURSOR.update` with the current key set to a pinned page, successfully update that row without doing a search, and then call `__wt_key_return`, `WT_CURSTD_KEY_INT` will be set, and the cursor will reference the correct on-page key (which hasn't changed).

What's worse, we can't "update" the key, even if we wanted to do the extra work, because absent a search we don't have the information we need to set the key.

That's why the key return code looks different from value return code, the value return code doesn't have the "it was set and hasn't changed" case.